server/src/main/java/org/jboss/as/server/Main.java

```diff
@@ -392,6 +388,12 @@ public static ServerEnvironmentWrapper determineEnvironment(String[] args, Prope
                     } else {
                         supplementalConfiguration = arg.substring(idx + 1);
                     }
+                } else if (CommandLineConstants.VERSION.equals(arg) || CommandLineConstants.SHORT_VERSION.equals(arg)
+                        || CommandLineConstants.OLD_VERSION.equals(arg) || CommandLineConstants.OLD_SHORT_VERSION.equals(arg)) {
+                    // Re-create using updated properties
+                    productConfig = ProductConfig.fromFilesystemSlot(Module.getBootModuleLoader(), WildFlySecurityManager.getPropertyPrivileged(ServerEnvironment.HOME_DIR, null), systemProperties);
+                    STDOUT.println(productConfig.getPrettyVersionString());
+                    return new ServerEnvironmentWrapper(ServerEnvironmentWrapper.ServerEnvironmentStatus.NORMAL);
```

Comment:
@soul2zimate @pferraro This tells me the handling of productConfig in this loop is wrong in general. This fixes it for -v but not for anything else.

The -v handling can remain at the top. It should so the change that fixes this doesn't imply it's needed to move it.

I think we need to add logic to all the else blocks that process system properties (-p and -D blocks) that check if HOME_DIR is set and if so recreate productConfig.

Probably leave it as null until that happens (or if HOME_DIR is known where we assign productConfig now), and the CommandLineUsageImpl stuff deals with a null ProductConfig by not adding the --stability arg to the output (i.e. default to it not being available until we know whether it is by loading the ProductConfig

Reply:
> I think we need to add logic to all the else blocks that process system properties (-p and -D blocks) that check if HOME_DIR is set and if so recreate productConfig.

Can we just move the handling of  system properties (-p and -D blocks) to the first blocks, and recreate `productConfig` if `jboss.home.dir` is found ?

Reply:
I misread your comment, don't bother with my previous comment via email(deleted).

I updated this PR, now it recreates `proudctConfig` if `jboss.home.dir` is found in system properties.